### PR TITLE
[Refactor] refactor hdfs scanner apppend_or_update column

### PR DESCRIPTION
**Pull request by dirtysalt**
## What I'm doing:

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/hdfs_scanner.cpp [hunk 1]
```diff
@@ -433,33 +425,21 @@ StatusOr<bool> HdfsScannerContext::should_skip_by_evaluating_not_existed_slots()
...
-                chunk_part_column->append_nulls(1);
-            } else {
-                chunk_part_column->append(*data_column, 0, 1);
-            }
-            chunk_part_column->assign(row_count, 0);
-        }
-
+        // create constant column here.
+        ColumnPtr chunk_part_column = partition_values[i]->clone_shared();
+        chunk_part_column->resize(row_count);
```
Comment (zombee0) [inline be/src/exec/hdfs_scanner.cpp:435]: const_null_column will has conflict with type in slot_desc

**Review by letian-jiang — APPROVED** [review]

File: be/src/exec/hdfs_scanner.cpp [hunk 2]
```diff
@@ -453,13 +440,9 @@ void HdfsScannerContext::append_or_update_partition_column_to_chunk(ChunkPtr* ch
...
         }
-
-        if (ck->is_slot_exist(slot_desc->id())) {
-            ck->update_column(std::move(chunk_part_column), slot_desc->id());
-        } else {
-            ck->append_column(std::move(chunk_part_column), slot_desc->id());
-        }
+        ck->append_or_update_column(std::move(chunk_part_column), slot_desc->id());
     }
+    ck->set_num_rows(row_count);
```
Comment (zombee0) [inline be/src/exec/hdfs_scanner.cpp:445]: do we need this?

**Review by zombee0 — COMMENTED** [review]
Left LGTM

File: be/src/exec/hdfs_scanner.cpp (hunk 2 above)
Comment (dirtysalt) [reply be/src/exec/hdfs_scanner.cpp:445]: not necessarily, but I guess there is no harm to do that.

**Review by sduzh — APPROVED** [review]

**Review by satanson — APPROVED** [review]

**Comment by dirtysalt** [comment]
@mergify backport branch-3.1

**Comment by dirtysalt** [comment]
@mergify backport branch-3.0

**Comment by dirtysalt** [comment]
@mergify backport branch-2.5

**Pull request merged**